### ISSUE TITLE
Assertion error in overwrite placeholder

### ISSUE THREAD

**Issue by matklad** [salsa-rs/salsa#70]
I am seeing an assertion error in salsa's from this code:

https://github.com/rust-analyzer/rust-analyzer/commit/3680ce50c0bdc3ac3ec9a3bafeaf6533e8dc9031

This snippet has **everything**:

 - it execute queries in parallel using rayon
 - it launches gc afterwards
 - it gets cancelled

This is the line that triggers assertion error:

https://github.com/salsa-rs/salsa/blob/v0.6.2/src/derived.rs#L537

Here's what I see in the logs:

<details>

```
ERROR [ra_analysis::imp] start RootDatabase { runtime: Runtime { id: RuntimeId { counter: 10 }, revision: R26902 } }
ERROR [ra_analysis::imp] files, symbols = 13450 13450
ERROR [ra_analysis::imp] end RootDatabase { runtime: Runtime { id: RuntimeId { counter: 10 }, revision: R26903 } }
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: RecvError', libcore/result.rs:1009:5
thread '<unnamed>' panicked at 'assertion failed: `(left == right)`
  left: `RuntimeId { counter: 279 }`,
 right: `RuntimeId { counter: 281 }`', /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/salsa-0.6.2/src/derived.rs:537:21
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::print
             at libstd/sys_common/backtrace.rs:71
             at libstd/sys_common/backtrace.rs:59
   2: std::panicking::default_hook::{{closure}}
             at libstd/panicking.rs:211
   3: std::panicking::default_hook
             at libstd/panicking.rs:227
   4: std::panicking::rust_panic_with_hook
             at libstd/panicking.rs:477
   5: std::panicking::continue_panic_fmt
             at libstd/panicking.rs:391
   6: rust_begin_unwind
             at libstd/panicking.rs:326
   7: core::panicking::panic_fmt
             at libcore/panicking.rs:77
   8: core::result::unwrap_failed
             at libcore/macros.rs:26
   9: <salsa::derived::DerivedStorage<DB, Q, MP>>::read_upgrade
             at libcore/result.rs:808
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/salsa-0.6.2/src/derived.rs:457
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/salsa-0.6.2/src/derived.rs:285
  10: <salsa::derived::DerivedStorage<DB, Q, MP> as salsa::plumbing::QueryStorageOps<DB, Q>>::try_fetch
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/salsa-0.6.2/src/derived.rs:266
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/salsa-0.6.2/src/derived.rs:612
  11: <salsa::QueryTable<'_, DB, Q>>::get
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/salsa-0.6.2/src/lib.rs:131
  12: ra_analysis::Analysis::on_enter
             at /home/matklad/projects/rust-analyzer/<::salsa::query_group macros>:25
             at crates/ra_analysis/src/imp.rs:178
             at crates/ra_analysis/src/lib.rs:224
  13: ra_lsp_server::main_loop::handlers::handle_on_enter
             at crates/ra_lsp_server/src/main_loop/handlers.rs:91
  14: std::panicking::try::do_call
             at crates/ra_lsp_server/src/main_loop/mod.rs:374
             at libstd/panic.rs:313
             at libstd/panicking.rs:310
  15: __rust_maybe_catch_panic
             at libpanic_unwind/lib.rs:103
  16: <rayon_core::job::HeapJob<BODY> as rayon_core::job::Job>::execute
             at libstd/panicking.rs:289
             at libstd/panic.rs:392
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-core-1.4.1/src/unwind.rs:18
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-core-1.4.1/src/spawn/mod.rs:72
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-core-1.4.1/src/job.rs:156
  17: rayon_core::registry::WorkerThread::wait_until_cold
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-core-1.4.1/src/job.rs:60
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-core-1.4.1/src/registry.rs:584
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-core-1.4.1/src/registry.rs:568
  18: rayon_core::registry::main_loop
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-core-1.4.1/src/registry.rs:544
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-core-1.4.1/src/registry.rs:666
  19: std::panicking::try::do_call
             at libstd/thread/mod.rs:409
             at libstd/panic.rs:313
             at libstd/panicking.rs:310
  20: __rust_maybe_catch_panic
             at libpanic_unwind/lib.rs:103
  21: <F as alloc::boxed::FnBox<A>>::call_box
             at libstd/panicking.rs:289
             at libstd/panic.rs:392
             at libstd/thread/mod.rs:408
             at liballoc/boxed.rs:646
  22: std::sys_common::thread::start_thread
             at liballoc/boxed.rs:656
             at libstd/sys_common/thread.rs:24
  23: std::sys::unix::thread::Thread::new::thread_start
             at libstd/sys/unix/thread.rs:90
  24: start_thread
  25: __GI___clone
ERROR [ra_lsp_server::main_loop] thread panicked :(
stack backtrace:
   0: ERRORstd [::ra_analysis::impsys] ::start unix::RootDatabasebacktrace {:: tracingruntime::: impRuntime:: {unwind_backtrace 
id: RuntimeId            {   at counter: libstd/sys/unix/backtrace/tracing/gcc_s.rs283: }49,
 revision  :  R126903 }:  }
std::sys_common::backtrace::print
             at libstd/sys_common/backtrace.rs:71
             at libstd/sys_common/backtrace.rs:59
   2: std::panicking::default_hook::{{closure}}
             at libstd/panicking.rs:211
   3: std::panicking::default_hook
             at libstd/panicking.rs:227
   4: std::panicking::rust_panic_with_hook
             at libstd/panicking.rs:477
   5: std::panicking::continue_panic_fmt
             at libstd/panicking.rs:391
   6: std::panicking::begin_panic_fmt
             at libstd/panicking.rs:346
   7: <salsa::derived::DerivedStorage<DB, Q, MP>>::overwrite_placeholder
             at /home/matklad/projects/rust-analyzer/<::std::macros::panic macros>:8
   8: <salsa::derived::DerivedStorage<DB, Q, MP>>::read_upgrade
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/salsa-0.6.2/src/derived.rs:397
   9: <salsa::derived::DerivedStorage<DB, Q, MP> as salsa::plumbing::QueryStorageOps<DB, Q>>::try_fetch
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/salsa-0.6.2/src/derived.rs:266
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/salsa-0.6.2/src/derived.rs:612
  10: <salsa::QueryTable<'_, DB, Q>>::get
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/salsa-0.6.2/src/lib.rs:131
  11: ra_analysis::Analysis::folding_ranges
             at /home/matklad/projects/rust-analyzer/<::salsa::query_group macros>:25
             at crates/ra_analysis/src/imp.rs:178
             at crates/ra_analysis/src/lib.rs:242
  12: ra_lsp_server::main_loop::handlers::handle_folding_range
             at crates/ra_lsp_server/src/main_loop/handlers.rs:422
  13: std::panicking::try::do_call
             at crates/ra_lsp_server/src/main_loop/mod.rs:374
             at libstd/panic.rs:313
             at libstd/panicking.rs:310
  14: __rust_maybe_catch_panic
             at libpanic_unwind/lib.rs:103
  15: <rayon_core::job::HeapJob<BODY> as rayon_core::job::Job>::execute
             at libstd/panicking.rs:289
             at libstd/panic.rs:392
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-core-1.4.1/src/unwind.rs:18
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-core-1.4.1/src/spawn/mod.rs:72
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-core-1.4.1/src/job.rs:156
  16: rayon_core::registry::WorkerThread::wait_until_cold
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-core-1.4.1/src/job.rs:60
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-core-1.4.1/src/registry.rs:584
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-core-1.4.1/src/registry.rs:568
  17: rayon_core::registry::main_loop
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-core-1.4.1/src/registry.rs:544
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-core-1.4.1/src/registry.rs:666
  18: std::panicking::try::do_call
             at libstd/thread/mod.rs:409
             at libstd/panic.rs:313
             at libstd/panicking.rs:310
  19: __rust_maybe_catch_panic
             at libpanic_unwind/lib.rs:103
  20: <F as alloc::boxed::FnBox<A>>::call_box
             at libstd/panicking.rs:289
             at libstd/panic.rs:392
             at libstd/thread/mod.rs:408
             at liballoc/boxed.rs:646
  21: std::sys_common::thread::start_thread
             at liballoc/boxed.rs:656
             at libstd/sys_common/thread.rs:24
  22: std::sys::unix::thread::Thread::new::thread_start
             at libstd/sys/unix/thread.rs:90
  23: start_thread
  24: __GI___clone
ERROR [ra_lsp_server::main_loop] thread 'thread panicked :(<unnamed>
' panicked at 'expected in-progress state', /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/salsa-0.6.2/src/derived.rs:551:22
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::print
             at libstd/sys_common/backtrace.rs:71
             at libstd/sys_common/backtrace.rs:59
   2: std::panicking::default_hook::{{closure}}
             at libstd/panicking.rs:211
   3: std::panicking::default_hook
             at libstd/panicking.rs:227
   4: std::panicking::rust_panic_with_hook
             at libstd/panicking.rs:477
   5: std::panicking::begin_panic
             at libstd/panicking.rs:411
   6: <salsa::derived::DerivedStorage<DB, Q, MP>>::overwrite_placeholder
             at /home/matklad/projects/rust-analyzer/<::std::macros::panic macros>:3
   7: <salsa::derived::DerivedStorage<DB, Q, MP>>::read_upgrade
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/salsa-0.6.2/src/derived.rs:397
   8: <salsa::derived::DerivedStorage<DB, Q, MP> as salsa::plumbing::QueryStorageOps<DB, Q>>::try_fetch
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/salsa-0.6.2/src/derived.rs:266
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/salsa-0.6.2/src/derived.rs:612
   9: ra_analysis::db::SyntaxDatabase::file_syntax
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/salsa-0.6.2/src/lib.rs:131
             at /home/matklad/projects/rust-analyzer/<::salsa::query_group macros>:25
  10: ra_analysis::descriptors::module::imp::submodules
             at crates/ra_analysis/src/descriptors/module/imp.rs:22
  11: <salsa::runtime::Runtime<DB>>::execute_query_implementation
             at /home/matklad/projects/rust-analyzer/<::salsa::query_group macros>:58
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/salsa-0.6.2/src/derived.rs:338
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/salsa-0.6.2/src/runtime.rs:255
  12: <salsa::derived::DerivedStorage<DB, Q, MP>>::read_upgrade
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/salsa-0.6.2/src/derived.rs:331
  13: <salsa::derived::DerivedStorage<DB, Q, MP> as salsa::plumbing::QueryStorageOps<DB, Q>>::maybe_changed_since
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/salsa-0.6.2/src/derived.rs:703
  14: <salsa::derived::DerivedStorage<DB, Q, MP>>::read_upgrade
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/salsa-0.6.2/src/derived.rs:900
             at libcore/iter/mod.rs:1506
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/salsa-0.6.2/src/derived.rs:898
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/salsa-0.6.2/src/derived.rs:310
  15: <salsa::derived::DerivedStorage<DB, Q, MP> as salsa::plumbing::QueryStorageOps<DB, Q>>::try_fetch
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/salsa-0.6.2/src/derived.rs:266
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/salsa-0.6.2/src/derived.rs:612
  16: ra_analysis::descriptors::module::ModulesDatabase::module_tree
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/salsa-0.6.2/src/lib.rs:131
             at /home/matklad/projects/rust-analyzer/<::salsa::query_group macros>:25
  17: ra_analysis::imp::AnalysisImpl::diagnostics
             at crates/ra_analysis/src/imp.rs:206
             at crates/ra_analysis/src/imp.rs:349
  18: ra_analysis::Analysis::diagnostics
             at crates/ra_analysis/src/lib.rs:283
  19: ra_lsp_server::main_loop::handlers::publish_diagnostics
             at crates/ra_lsp_server/src/main_loop/handlers.rs:594
  20: std::panicking::try::do_call
             at crates/ra_lsp_server/src/main_loop/mod.rs:410
             at libstd/panic.rs:313
             at libstd/panicking.rs:310
  21: __rust_maybe_catch_panic
             at libpanic_unwind/lib.rs:103
  22: <rayon_core::job::HeapJob<BODY> as rayon_core::job::Job>::execute
             at libstd/panicking.rs:289
             at libstd/panic.rs:392
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-core-1.4.1/src/unwind.rs:18
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-core-1.4.1/src/spawn/mod.rs:72
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-core-1.4.1/src/job.rs:156
  23: rayon_core::registry::WorkerThread::wait_until_cold
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-core-1.4.1/src/job.rs:60
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-core-1.4.1/src/registry.rs:584
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-core-1.4.1/src/registry.rs:568
  24: rayon_core::registry::main_loop
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-core-1.4.1/src/registry.rs:544
             at /home/matklad/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-core-1.4.1/src/registry.rs:666
  25: std::panicking::try::do_call
             at libstd/thread/mod.rs:409
             at libstd/panic.rs:313
             at libstd/panicking.rs:310
  26: __rust_maybe_catch_panic
             at libpanic_unwind/lib.rs:103
  27: <F as alloc::boxed::FnBox<A>>::call_box
             at libstd/panicking.rs:289
             at libstd/panic.rs:392
             at libstd/thread/mod.rs:408
             at liballoc/boxed.rs:646
  28: std::sys_common::thread::start_thread
             at liballoc/boxed.rs:656
             at libstd/sys_common/thread.rs:24
  29: std::sys::unix::thread::Thread::new::thread_start
             at libstd/sys/unix/thread.rs:90
  30: start_thread
  31: __GI___clone
ERROR [ra_lsp_server::main_loop] thread panicked :(
ERROR [ra_analysis::imp] files, symbols = 13450 13450
ERROR [ra_analysis::imp] end RootDatabase { runtime: Runtime { id: RuntimeId { counter: 283 }, revision: R26903 } }
```

<details>

**Comment by matklad** [comment]
I've reviewed the code which could be responsible for this:

* gc
* PanicGuard
* read_upgrade

I don't see any obvious problem. The only explanation I can come up with is that my own code panics, silently (there are no analyzer-related panics in the log), and that causes "PanicGuard is not multithreaded" problem.  However, I can't explain the "silent" bit. Will try to do some kind of reproduction when I have more time.

**Comment by matklad** [comment]
Steps to reproduce:

* install vscode
* clone https://github.com/rust-analyzer/rust-analyzer at 3680ce5
* run `cargo install-code`
* inside rust-lang/rust project (it just has a lot of files) run `code .`

**Comment by matklad** [comment]
link to zulip convo: 

https://salsa.zulipchat.com/#narrow/stream/145099-general/subject/GC.20battle.20tested/near/136706075

TL;DR: GC + [this line](https://github.com/salsa-rs/salsa/blob/c018196ffbb6f038a582cc93812099076a541f0b/src/derived.rs#L738) probably trigger a bug.
